File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_12/java/Solution.java
package com.hi.dhl.algorithms.offer._12.java;

/**
 * <pre>
 *     author: dhl
 *     desc  :
 * </pre>
 */

class Solution {
    public boolean exist(char[][] board, String word) {
        if (board.length <= 0 || board[0].length <= 0) {
            return false;
        }
        char[] words = word.toCharArray();
        int row = board.length;
        int colum = board[0].length;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < colum; j++) {
                if (dfs(board, words, i, j, 0, row, colum)) {
                    return true;
                }
            }
        }
        return false;
    }

    boolean dfs(char[][] borad, char[] words, int i, int j, int k, int row, int colum) {
        if (i > row - 1 || i < 0 || j > colum - 1 || j < 0 || borad[i][j] != words[k]) {
            return false;
        }

        if (k == words.length - 1) {
            return true;
        }

        char tmp = borad[i][j];
        borad[i][j] = '/';
        boolean result = dfs(borad, words, i + 1, j, k + 1, row, colum) || dfs(borad, words, i - 1, j, k + 1, row, colum) || dfs(borad, words, i, j + 1, k + 1, row, colum) || dfs(borad, words, i, j - 1, k + 1, row, colum);
        borad[i][j] = tmp;
        return result;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/offer/_12/kotlin/Solution.kt
package com.hi.dhl.algorithms.offer._12.kotlin

/**
 * <pre>
 *     author: dhl
 *     desc  :
 * </pre>
 */

class Solution {
    fun exist(board: Array<CharArray>, word: String): Boolean {
        if (board.size <= 0 || board[0].size <= 0) {
            return false
        }
        val row = board.size
        val colum = board[0].size
        val words = word.toCharArray()
        for (i in 0 until row) {
            for (j in 0 until colum) {
                if (dfs(board, words, i, j, 0, row, colum)) {
                    return true
                }
            }
        }
        return false
    }

    fun dfs(board: Array<CharArray>, words: CharArray, i: Int, j: Int, k: Int, row: Int, colum: Int): Boolean {
        if (i > row - 1 || i < 0 || j > colum - 1 || j < 0 || board[i][j] != words[k]) {
            return false
        }

        if (k == words.size - 1) {
            return true
        }
        val tmp = board[i][j]
        board[i][j] = '/'
        val result =
            dfs(board, words, i + 1, j, k + 1, row, colum) || dfs(board, words, i - 1, j, k + 1, row, colum) || dfs(
                board,
                words,
                i,
                j + 1,
                k + 1,
                row,
                colum
            ) || dfs(board, words, i, j - 1, k + 1, row, colum)
        board[i][j] = tmp
        return result
    }
}